File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第8章：边缘检测/8.6/canny.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>
using namespace cv;
//非极大值抑制
Mat non_maximum_suppression_default(Mat dx, Mat dy)
{	//使用平方和开方的方式计算边缘强度
	Mat edgeMag;
	cv::magnitude(dx, dy, edgeMag);
	//宽高
	int rows = dx.rows;
	int cols = dy.cols;
	//边缘强度的非极大抑制
	Mat edgeMag_nonMaxSup = Mat::zeros(dx.size(), dx.type());
	for (int r = 1; r < rows-1; r++)
	{
		for (int c = 1; c < cols-1; c++)
		{
			float x = dx.at<float>(r, c);
			float y = dy.at<float>(r, c);
			//梯度方向
			float angle = atan2f(y, x)/CV_PI*180;
			//当前位置的边缘强度
			float mag = edgeMag.at<float>(r, c);
			// 左 / 右方向比较
			if (abs(angle) < 22.5 || abs(angle) > 157.5)
			{
				float left = edgeMag.at<float>(r, c - 1);
				float right = edgeMag.at<float>(r, c + 1);
				if (mag > left && mag > right)
					edgeMag_nonMaxSup.at<float>(r, c) = mag;
			}
			//左上 / 右下方向比较
			if ((angle >= 22.5 && angle < 67.5) || (angle < -112.5 && angle >= 157.5))
			{
				float leftTop = edgeMag.at<float>(r - 1, c - 1);
				float rightBottom = edgeMag.at<float>(r + 1, c + 1);
				if (mag > leftTop && mag > rightBottom)
					edgeMag_nonMaxSup.at<float>(r, c) = mag;
			}
			//上 / 下方向比较
			if ((angle >= 67.5 && angle <= 112.5) || (angle >= -112.5 && angle <= -67.5))
			{
				float top = edgeMag.at<float>(r - 1, c);
				float bottom = edgeMag.at<float>(r + 1, c);
				if (mag > top && mag > bottom)
					edgeMag_nonMaxSup.at<float>(r, c) = mag;
			}
			//右上 / 左下方向
			if ((angle > 112.5 && angle <= 157.5) || (angle > -67.5 && angle <= -22.5))
			{
				float rightTop = edgeMag.at<float>(r - 1, c + 1);
				float leftBottom = edgeMag.at<float>(r + 1, c - 1);
				if (mag > rightTop && mag > leftBottom)
					edgeMag_nonMaxSup.at<float>(r, c) = mag;
			}
		}
	}
	return edgeMag_nonMaxSup;
}
//非极大值抑制 ： 插值比较
Mat non_maximum_suppression_Inter(Mat dx, Mat dy)
{
	//使用平方和开方的方式计算边缘强度
	Mat edgeMag;
	cv::magnitude(dx, dy, edgeMag);
	//宽高
	int rows = dx.rows;
	int cols = dy.cols;
	//边缘强度的非极大抑制
	Mat edgeMag_nonMaxSup = Mat::zeros(dx.size(), dx.type());
	for (int r = 1; r < rows-1; r++)
	{
		for (int c = 1; c < cols-1; c++)
		{
			float x = dx.at<float>(r, c);
			float y = dy.at<float>(r, c);
			if (x == 0 && y == 0)
				continue;
			float angle = atan2f(y, x) / CV_PI * 180;
			//领域内八个方向上的边缘强度
			float leftTop = edgeMag.at<float>(r - 1, c - 1);
			float top = edgeMag.at<float>(r - 1, c);
			float rightBottom = edgeMag.at<float>(r + 1, c + 1);
			float right = edgeMag.at<float>(r, c+1);
			float rightTop = edgeMag.at<float>(r - 1, c + 1);
			float leftBottom = edgeMag.at<float>(r + 1, c - 1);
			float bottom = edgeMag.at<float>(r + 1, c);
			float left = edgeMag.at<float>(r, c - 1);
			float mag = edgeMag.at<float>(r, c);
			//左上方与上方的插值 右下方和下方的插值
			if ((angle > 45 && angle <= 90) || (angle > -135 && angle <= -90))
			{
				float ratio = x / y;
				float top = edgeMag.at<float>(r - 1, c);
				//插值
				float leftTop_top = ratio*leftTop + (1 - ratio)*top;
				float rightBottom_bottom = ratio*rightBottom + (1 - ratio)*bottom;
				if (mag > leftTop_top && mag > rightBottom_bottom)
					edgeMag_nonMaxSup.at<float>(r, c) = mag;
			}
			//右上方和上方的插值 左下方和下方的插值
			if ((angle > 90 && angle <= 135) || (angle > -90 && angle <= -45))
			{
				float ratio = abs(x / y);
				float rightTop_top = ratio*rightTop + (1 - ratio)*top;
				float leftBottom_bottom = ratio*leftBottom + (1 - ratio)*bottom;
				if (mag > rightTop_top && mag > leftBottom_bottom)
					edgeMag_nonMaxSup.at<float>(r, c) = mag;
			}
			//左上方和左方的插值 右下方和右方的插值
			if ((angle >= 0 && angle <= 45) || (angle > -180 && angle <= -135))
			{
				float ratio = y / x;
				float rightBottom_right = ratio*rightBottom + (1 - ratio)*right;
				float leftTop_left = ratio*leftTop + (1 - ratio)*left;
				if (mag > rightBottom_right && mag > leftTop_left)
					edgeMag_nonMaxSup.at<float>(r, c) = mag;
			}
			//右上方和右方的插值 左下方和左方的插值
			if ((angle > 135 && angle <= 180) || (angle > -45 && angle <= 0))
			{
				float ratio = abs(y / x);
				float rightTop_right = ratio*rightTop + (1 - ratio)*right;
				float leftBottom_left = ratio*leftBottom + (1 - ratio)*left;
				if (mag > rightTop_right && mag > leftBottom_left)
					edgeMag_nonMaxSup.at<float>(r, c) = mag;
			}
		}
	}
	return edgeMag_nonMaxSup;
}
//确定一个点的坐标是否在图像范围内
bool checkInRange(int r, int c, int rows, int cols)
{
	if (r >= 0 && r < rows && c>=0 && c < cols)
		return true;
	else
		return false;
}
//从确定边缘点出发，延长边缘
void trace(Mat edgeMag_nonMaxSup, Mat &edge, float lowerThresh, int r, int c, int rows,int cols)
{
	if (edge.at<uchar>(r, c) == 0)
	{
		edge.at<uchar>(r, c) = 255;
		for (int i = -1; i <= 1; i++)
		{
			for (int j = -1; j <= 1; j++)
			{
				float mag = edgeMag_nonMaxSup.at<float>(r+i, c+j);
				if (checkInRange(r + i, c + j, rows, cols) && mag >= lowerThresh)
					trace(edgeMag_nonMaxSup, edge, lowerThresh, r+i, c+j, rows, cols);
			}
		}
	}
}
//双阈值的滞后阈值处理
Mat hysteresisThreshold(Mat edgeMag_nonMaxSup, float lowerThresh, float upperThresh)
{	
	//宽高
	int rows = edgeMag_nonMaxSup.rows;
	int cols = edgeMag_nonMaxSup.cols;
	//最后的边缘输出图
	Mat edge = Mat::zeros(Size(cols, rows), CV_8UC1);
	//滞后阈值处理
	for (int r = 1; r < rows - 1; r++)
	{
		for (int c = 1; c < cols-1; c++)
		{
			float mag = edgeMag_nonMaxSup.at<float>(r, c);
			//大于高阈值，可作为确定边缘点被接受
			//并以该点位起始点延长边缘
			if(mag>=upperThresh)
				trace(edgeMag_nonMaxSup, edge, lowerThresh, r, c, rows, cols);
			//小于低阈值的直接被剔除
			if (mag < lowerThresh)
				edge.at<uchar>(r, c) = 0;
		}
	}
	return  edge;
}
int main(int argc,char*argv[])
{
	//输入图像
	Mat image = imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	if (!image.data)
		return -1;
	//第一步：sobel 卷积，这里采用 OpenCV 封装的 Sobel 函数
	Mat dx, dy;
	cv::Sobel(image, dx, CV_32FC1, 1, 0, 3);
	cv::Sobel(image, dy, CV_32FC1, 0, 1, 3);
	//边缘强度的灰度级显示
	Mat edgeMag;
	cv::magnitude(dx, dy, edgeMag);
	Mat mag;
	edgeMag.convertTo(mag, CV_8UC1);
	cv::imshow("边缘强度的灰度级显示", mag);
	cv::imwrite("mag.jpg", mag);
	//第二步：非极大值抑制
	Mat edgeMag_nonMaxSup = non_maximum_suppression_default(dx, dy);
	//显示非极大值抑制后的边缘强度图
	Mat nonMaxSup;
	edgeMag_nonMaxSup.convertTo(nonMaxSup, CV_8UC1);
	cv::imshow("非极大值抑制", nonMaxSup);
	cv::imwrite("nonMaxSup.jpg", nonMaxSup);
	//第三步：双阈值的滞后阈值处理
	float lowerThresh = 40;
	float upperThresh = 150;
	Mat edge = hysteresisThreshold(edgeMag_nonMaxSup, lowerThresh, upperThresh);
	imshow("CannyEdge", edge);
	cv::imwrite("CannyEdge.jpg", edge);
	//与单阈值处理进行比较
	Mat lowerThreshEdge;
	cv::threshold(edgeMag_nonMaxSup, lowerThreshEdge, lowerThresh, 255, CV_THRESH_BINARY);
	cv::convertScaleAbs(lowerThreshEdge, lowerThreshEdge, 1.0, 0);
	cv::imshow("低阈值处理后的效果", lowerThreshEdge);
	cv::imwrite("lowerThreshEdge.jpg", lowerThreshEdge);
	Mat upperThreshEdge;
	cv::threshold(edgeMag_nonMaxSup, upperThreshEdge, upperThresh, 255, CV_THRESH_BINARY);
	cv::convertScaleAbs(upperThreshEdge, upperThreshEdge, 1.0, 0.0);
	cv::imshow("高阈值处理后的效果", upperThreshEdge);
	cv::imwrite("upperThreshEdge.jpg", upperThreshEdge);
	waitKey(0);
	return 0;
}

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第8章：边缘检测/8.6/canny.py
# -*- coding: utf-8 -*-
import numpy as np
import sys
import math
import cv2
import sobel #注意sobel边缘检测
#边缘检测
#非极大值抑制
def non_maximum_suppression_default(dx,dy):
    #边缘强度
    edgeMag = np.sqrt(np.power(dx,2.0) + np.power(dy,2.0))
    #宽、高
    rows,cols = dx.shape
    #梯度方向
    gradientDirection = np.zeros(dx.shape)
    #边缘强度非极大值抑制
    edgeMag_nonMaxSup = np.zeros(dx.shape)
    for r in range(1,rows-1):
        for c in range(1,cols-1):
            #angle 的范围 [0,180] [-180,0]
            angle = math.atan2(dy[r][c],dx[r][c])/math.pi*180
            gradientDirection[r][c] = angle
            #左 / 右方向
            if(abs(angle)<22.5 or abs(angle) >157.5):
                if(edgeMag[r][c]>edgeMag[r][c-1] and edgeMag[r][c] > edgeMag[r][c+1]):
                    edgeMag_nonMaxSup[r][c] = edgeMag[r][c]
            #左上 / 右下方向  
            if(angle>=22.5 and angle < 67.5 or(-angle > 112.5 and -angle <= 157.5)):
                if(edgeMag[r][c] > edgeMag[r-1][c-1] and edgeMag[r][c]>edgeMag[r+1][c+1]):
                     edgeMag_nonMaxSup[r][c] = edgeMag[r][c]
            #上 / 下方向
            if((angle>=67.5 and angle<=112.5) or (angle>=-112.5 and angle<=-67.5)):
                if(edgeMag[r][c] > edgeMag[r-1][c] and edgeMag[r][c] > edgeMag[r+1][c]):
                    edgeMag_nonMaxSup[r][c] = edgeMag[r][c]
            #右上 / 左下方向
            if((angle>112.5 and angle<=157.5) or(-angle>=22.5 and -angle< 67.5 )):
                if(edgeMag[r][c]>edgeMag[r-1][c+1] and edgeMag[r][c] > edgeMag[r+1][c-1]):
                    edgeMag_nonMaxSup[r][c] = edgeMag[r][c]
    return edgeMag_nonMaxSup
#非极大值抑制：插值比较
def non_maximum_suppression_Inter(dx,dy):
    #边缘强度
    edgeMag = np.sqrt(np.power(dx,2.0)+np.power(dy,2.0))
    #宽、高
    rows,cols = dx.shape
    #梯度方向
    gradientDirection = np.zeros(dx.shape)
    #边缘强度的非极大值抑制
    edgeMag_nonMaxSup = np.zeros(dx.shape)
    for r in range(1,rows-1):
        for c in range(1,cols-1):
            if dy[r][c] ==0 and dx[r][c] == 0:
                continue
            #angle的范围 [0,180],[-180,0]
            angle = math.atan2(dy[r][c],dx[r][c])/math.pi*180
            gradientDirection[r][c] = angle
            #左上方和上方的插值 右下方和下方的插值
            if (angle > 45 and angle <=90) or (angle > -135 and angle <=-90):
                ratio = dx[r][c]/dy[r][c]
                leftTop_top = ratio*edgeMag[r-1][c-1]+(1-ratio)*edgeMag[r-1][c]
                rightBottom_bottom = (1-ratio)*edgeMag[r+1][c] + ratio*edgeMag[r+1][c+1]
                if edgeMag[r][c] >  leftTop_top and edgeMag[r][c] > rightBottom_bottom:
                    edgeMag_nonMaxSup[r][c]  = edgeMag[r][c]
            #右上方和上方的插值 左下方和下方的插值
            if (angle>90 and angle<=135) or (angle>-90 and angle <= -45):
                ratio = abs(dx[r][c]/dy[r][c])
                rightTop_top = ratio*edgeMag[r-1][c+1] + (1-ratio)*edgeMag[r-1][c]
                leftBottom_bottom = ratio*edgeMag[r+1][c-1] + (1-ratio)*edgeMag[r+1][c]
                if edgeMag[r][c] > rightTop_top and edgeMag[r][c] > leftBottom_bottom:
                    edgeMag_nonMaxSup[r][c]  = edgeMag[r][c]
            #左上方和左方的插值 右下方和右方的插值
            if (angle>=0 and angle <=45) or (angle>-180 and angle <= -135):
                ratio = dy[r][c]/dx[r][c]
                rightBottom_right = ratio*edgeMag[r+1][c+1]+(1-ratio)*edgeMag[r][c+1]
                leftTop_left = ratio*edgeMag[r-1][c-1]+(1-ratio)*edgeMag[r][c-1]
                if edgeMag[r][c] > rightBottom_right and edgeMag[r][c] > leftTop_left:
                    edgeMag_nonMaxSup[r][c]  = edgeMag[r][c]
            #右上方和右方的插值 左下方和左方的插值
            if(angle>135 and angle<=180) or (angle>-45 and angle <=0):
                ratio = abs(dy[r][c]/dx[r][c])
                rightTop_right = ratio*edgeMag[r-1][c+1]+(1-ratio)*edgeMag[r][c+1]
                leftBottom_left = ratio*edgeMag[r+1][c-1]+(1-ratio)*edgeMag[r][c-1]
                if edgeMag[r][c] > rightTop_right and edgeMag[r][c] > leftBottom_left:
                    edgeMag_nonMaxSup[r][c]  = edgeMag[r][c]
    return edgeMag_nonMaxSup
#判断一个点的坐标是否在图像范围内
def checkInRange(r,c,rows,cols):
    if r>=0 and r<rows and c>=0 and c<cols:
        return True
    else:
        return False
def trace(edgeMag_nonMaxSup,edge,lowerThresh,r,c,rows,cols):
    #大于阈值为确定边缘点
    if edge[r][c] == 0:
        edge[r][c]=255
        for i in range(-1,2):
            for j in range(-1,2):
                if checkInRange(r+i,c+j,rows,cols) and edgeMag_nonMaxSup[r+i][c+j] >= lowerThresh:
                    trace(edgeMag_nonMaxSup,edge,lowerThresh,r+i,c+j,rows,cols)
#滞后阈值
def hysteresisThreshold(edge_nonMaxSup,lowerThresh,upperThresh):
    #宽高
    rows,cols = edge_nonMaxSup.shape
    edge = np.zeros(edge_nonMaxSup.shape,np.uint8)
    for r in range(1,rows-1):
        for c in range(1,cols-1):
            #大于高阈值，设置为确定边缘点，而且以该点为起始点延长边缘
            if edge_nonMaxSup[r][c] >= upperThresh:
                trace(edgeMag_nonMaxSup,edge,lowerThresh,r,c,rows,cols)
            #小于低阈值，被剔除
            if edge_nonMaxSup[r][c]< lowerThresh:
                edge[r][c] = 0
    return edge
#主函数
if __name__ =="__main__":
    if len(sys.argv)>1:
        image = cv2.imread(sys.argv[1],cv2.CV_LOAD_IMAGE_GRAYSCALE)
    else:
        print "Usge:python canny.py imageFile"
    # ------- canny 边缘检测 -----------
    #第一步： 基于 sobel 核的卷积
    image_sobel_x,image_sobel_y = sobel.sobel(image,3)
    #边缘强度：两个卷积结果对应位置的平方和
    edge = np.sqrt(np.power(image_sobel_x,2.0) + np.power(image_sobel_y,2.0))
    #边缘强度的灰度级显示
    edge[edge>255] = 255
    edge = edge.astype(np.uint8)
    cv2.imshow("sobel edge",edge)
    #第二步：非极大值抑制
    edgeMag_nonMaxSup = non_maximum_suppression_default(image_sobel_x,image_sobel_y)
    edgeMag_nonMaxSup[edgeMag_nonMaxSup>255] =255
    edgeMag_nonMaxSup = edgeMag_nonMaxSup.astype(np.uint8)
    cv2.imshow("edgeMag_nonMaxSup",edgeMag_nonMaxSup)
    #第三步：双阈值滞后阈值处理，得到 canny 边缘
    #滞后阈值的目的就是最后决定处于高阈值和低阈值之间的是否为边缘点
    edge = hysteresisThreshold(edgeMag_nonMaxSup,60,180)
    lowerThresh = 40
    upperThresh = 150
    cv2.imshow("canny",edge)
    cv2.imwrite("canny.jpg",edge)
    # -------以下是为了单阈值与滞后阈值的结果比较 ------
    #大于高阈值 设置为白色 为确定边缘
    EDGE = 255
    #小于低阈值的设置为黑色 表示不是边缘，被剔除
    NOEDGE = 0
    #而大于等于低阈值 小于高阈值的设置为灰色，标记为可能的边缘
    POSSIBLE_EDGE = 128
    tempEdge = np.copy(edgeMag_nonMaxSup)
    rows,cols = tempEdge.shape
    for r in xrange(rows):
        for c in xrange(cols):
            if tempEdge[r][c]>=upperThresh:
                tempEdge[r][c] = EDGE
            elif tempEdge[r][c]<lowerThresh:
                tempEdge[r][c] = NOEDGE
            else:
                tempEdge[r][c] = POSSIBLE_EDGE
    cv2.imshow("tempEdge",tempEdge)
    lowEdge = np.copy(edgeMag_nonMaxSup)
    lowEdge[lowEdge>60] = 255
    lowEdge[lowEdge<60] = 0
    cv2.imshow("lowEdge",lowEdge)
    upperEdge = np.copy(edgeMag_nonMaxSup)
    upperEdge[upperEdge>180]=255
    upperEdge[upperEdge<=180]=0
    cv2.imshow("upperEdge",upperEdge)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
        